packages/core/src/lib/markdown-it/highlight/helper.ts

```diff
@@ -6,3 +6,61 @@ export function splitCodeAndIndentation(codeStr: string) {
   const content = codeStr.substring(codeStartIdx);
   return [indents, content];
 }
+
+export enum BOUNDARY_TYPE {
+  Start,
+  End,
+}
+
+export interface Boundary {
+  index: number;
+  type: BOUNDARY_TYPE;
```

Comment:
Instead of having a `BOUNDARY_TYPE.Start` and `BOUNDARY_TYPE.End`, can we represent boundaries as a pair of integers `(start, end)` or have a Boundary type with `{ startIndex: number; endIndex: number }`?
It looks like we always add a Boundary with both a Start and an End, so representing it as a pair may help us maintain this invariant and reduce potential misrepresentations/bugs.

Reply:
If I am not wrong problem with this representation is that during collate interval I actually want to separate each bound instead of binding them as a pair for the method to work effectively in one n pass. Each interval is only closed when the number of ends equal to beginnings a bit like those uhm open closing braces parsing

Reply:
I see - if I'm understanding the purpose of collateAllIntervals correctly, I think merging the intervals can still be done in one pass with the paired boundaries, especially since you sort the array beforehand. 
Some possible solution in python that solves for [merging boundaries](https://leetcode.com/problems/merge-intervals/description/):
```python
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals = sorted(intervals)
        result = []
        prev_s, prev_e = intervals[0][0], intervals[0][1]
        for arr in intervals[1:]:
            s, e = arr[0], arr[1]
            if s > prev_e:
                # no overlap
                result.append([prev_s, prev_e])
                prev_s, prev_e = s, e
            else:
                # there is overlap
                prev_e = max(e, prev_e)
        result.append([prev_s, prev_e])
        return result
```

Reply:
I see, I can certainly adopt this approach. Thx.